Meta/Lagom/Tools/CodeGenerators/LibWeb/GenerateWebGLRenderingContext.cpp

```diff
@@ -487,6 +502,282 @@ static Vector<GLchar> null_terminated_string(StringView string)
+    return SkColorType::kUnknown_SkColorType;
+}
+)~~~");
+    } else {
+        implementation_file_generator.append(R"~~~(
+static constexpr Optional<int> opengl_format_number_of_components(WebIDL::UnsignedLong format)
+{
+    switch (format) {
+    case GL_LUMINANCE:
+    case GL_ALPHA:
+        return 1;
+    case GL_LUMINANCE_ALPHA:
+        return 2;
+    case GL_RGB:
+        return 3;
+    case GL_RGBA:
+        return 4;
+    default:
+        return OptionalNone {};
+    }
+}
+
+static constexpr Optional<int> opengl_type_size_in_bytes(WebIDL::UnsignedLong type)
+{
+    switch (type) {
+    case GL_UNSIGNED_BYTE:
+        return 1;
+    case GL_UNSIGNED_SHORT_5_6_5:
+    case GL_UNSIGNED_SHORT_4_4_4_4:
+    case GL_UNSIGNED_SHORT_5_5_5_1:
+        return 2;
+    default:
+        return OptionalNone {};
+    }
+}
+
+static constexpr SkColorType opengl_format_and_type_to_skia_color_type(WebIDL::UnsignedLong format, WebIDL::UnsignedLong type)
+{
+    switch (format)
+    {
+    case GL_RGB:
+        switch (type) {
+        case GL_UNSIGNED_BYTE:
+            return SkColorType::kRGB_888x_SkColorType;
+        case GL_UNSIGNED_SHORT_5_6_5:
+            return SkColorType::kRGB_565_SkColorType;
+        default:
+            break;
+        }
+        break;
+    case GL_RGBA:
+        switch (type) {
+        case GL_UNSIGNED_BYTE:
+            return SkColorType::kRGBA_8888_SkColorType;
+        case GL_UNSIGNED_SHORT_4_4_4_4:
+            // FIXME: This is not exactly the same as RGBA.
+            return SkColorType::kARGB_4444_SkColorType;
+        case GL_UNSIGNED_SHORT_5_5_5_1:
+            dbgln("WebGL FIXME: Support conversion to RGBA5551.");
+            break;
+        default:
+            break;
+        }
+        break;
+    case GL_ALPHA:
+        switch (type) {
+        case GL_UNSIGNED_BYTE:
+            return SkColorType::kAlpha_8_SkColorType;
+        default:
+            break;
+        }
+        break;
+    case GL_LUMINANCE:
+        switch (type) {
+        case GL_UNSIGNED_BYTE:
+            return SkColorType::kGray_8_SkColorType;
+        default:
+            break;
+        }
+        break;
+    default:
+        break;
+    }
+
+    dbgln("WebGL: Unsupported format and type combination. format: 0x{:04x}, type: 0x{:04x}", format, type);
+    return SkColorType::kUnknown_SkColorType;
+}
+)~~~");
+    }
+
+    implementation_file_generator.append(R"~~~(
+struct ConvertedTexture {
+    ByteBuffer buffer;
+    int width { 0 };
+    int height { 0 };
+};
+
+static Optional<ConvertedTexture> read_and_pixel_convert_texture_image_source(Variant<GC::Root<ImageBitmap>, GC::Root<ImageData>, GC::Root<HTMLImageElement>, GC::Root<HTMLCanvasElement>, GC::Root<HTMLVideoElement>> const& source, WebIDL::UnsignedLong format, WebIDL::UnsignedLong type, Optional<int> destination_width = OptionalNone {}, Optional<int> destination_height = OptionalNone {})
```

Comment:
can we define this function as a Gfx::Bitmap (or Gfx::ImmutableBitmap) method? it would be good to have as less code as possible inline code in generator since it's easier to read/edit with syntax highlighting.

Reply:
We definitely could do this, but let's do it in a follow-up :)